E2E test for Project creation

Shhhhhubh opened the pull request:
Addresses
- Complete flow test for the creation of a find type project along with firebase verification of groups and tasks.



## Changes

* Test script python file 
* Geojson file 
* json input/compare file
*  cover image
## This PR doesn't introduce any:

- [ ] temporary files, auto-generated files or secret keys
- [ ] n+1 queries
- [ ] flake8 issues
- [ ] `print`
- [ ] typos
- [ ] unwanted comments

## This PR contains valid:

- [ ] tests
- [ ] permission checks (tests here too)
- [ ] translations

babinkarmacharya79 committed:
test(project): create test script for project creation

Shhhhhubh committed:
test(project): use new status api

Shhhhhubh commented on test_data/project/create_find_project.json:
```diff
@@ -0,0 +1,146 @@
+        "projectInstruction": "temporary shelters",
+        "additionalInfoUrl": "https://reliefweb.int/disaster/fl-2025-000123-npl",
+        "description": "A project aimed at mapping key infrastructure and affected communities to support flood response operations in Nepal.",
+        "requestingOrganization":""
```
Let's define a name. We can create or fetch that organization using a factory later.

Shhhhhubh commented on apps/project/asset/image.jpg:
Let's move this to mapswipe-asset submodule

Shhhhhubh commented on apps/project/asset/find_geojson.geojson:
- Let's move this to mapswipe-asset submodule
- Let's format this
- Let's use json5

Shhhhhubh commented on test_data/project/create_find_project.json:
```diff
@@ -0,0 +1,146 @@
+        "clientId": "01K339HAWZCTCBRVQMQFWRTHXX",
+        "topic": "Emergency Response - Flood 2025",
+        "region": "South Asia - Nepal",
+        "projectNumber": 1024,
```
Why is this so large?

Shhhhhubh commented on test_data/project/create_find_project.json:
```diff
@@ -0,0 +1,146 @@
+        "region": "South Asia - Nepal",
+        "projectNumber": 1024,
+        "projectType": "FIND",
+        "projectInstruction": "temporary shelters",
```
This is lookFor field. Rename this field.


Please define a separate one for projectInstruction.
projectInstruction="Identify temporary shelters"

Shhhhhubh commented on test_data/project/create_find_project.json:
```diff
@@ -0,0 +1,146 @@
+        "verificationNumber": 10,
+        "projectInstruction": "temporary shelters",
+        "groupSize": 15,
+        "team":"",
```
Please use `null` to indicate absence of value.

Shhhhhubh commented on test_data/project/create_find_project.json:
```diff
@@ -0,0 +1,146 @@
+        "groupSize": 15,
+        "team":"",
+        "topic": "Emergency Response - Flood 2025",
+        "tutorial":"",
```
Please use `null` to indicate absence of value.

Shhhhhubh commented on test_data/project/create_find_project.json:
```diff
@@ -0,0 +1,146 @@
+        "tutorial":"",
+        "maxTasksPerUser": 20,
+        "projectNumber": 1024,
+        "region": "South Asia - Nepal",
```
```suggestion
        "region": "Nepal",
```

Shhhhhubh commented on test_data/project/create_find_project.json:
```diff
@@ -0,0 +1,146 @@
+            "find":{
+                "zoomLevel":18,
+                "tileServerProperty":{
+                    "name":"BING",
```
```suggestion
                    "name": "BING",
```

Shhhhhubh commented on test_data/project/create_find_project.json:
```diff
@@ -0,0 +1,146 @@
+        "maxTasksPerUser": 20,
+        "projectNumber": 1024,
+        "region": "South Asia - Nepal",
+        "requestingOrganization":"",
```
We probably don't need to send values that are already sent on create.

Shhhhhubh commented on test_data/project/create_find_project.json:
```diff
@@ -0,0 +1,146 @@
+        "projectNumber": 1024,
+        "region": "South Asia - Nepal",
+        "requestingOrganization":"",
+        "image": "",
```
Define the path for the image here.
The test should automatically create Cover Image and attach to the projectå

Shhhhhubh commented on test_data/project/create_find_project.json:
```diff
@@ -0,0 +1,146 @@
+                        "credits":"© 2019 Microsoft Corporation, Earthstar Geographics SIO"
+                    }
+                },
+                "aoiGeometry":""
```
Define the path for the aoi geometry here.
The test should automatically create AOI Geometry and attach to the projectå

Shhhhhubh commented on test_data/project/create_find_project.json:
```diff
@@ -0,0 +1,146 @@
+            }
+        }
+    },
+    "update_project_1_status":{
```
Put everything in an array.
The order defines the order or execution
Add a separate key to identify action.

Shhhhhubh commented on test_data/project/create_find_project.json:
```diff
@@ -0,0 +1,146 @@
+      "status":"PUBLISHED",
+      "clientId": "01K339HAWZCTCBRVQMQFWRTHXX"
+    },
+    "final_processed_project":{
+      "contributorCount":0,
```
the data for this section should be taken by creating the above project in our old platform (manager dashboard). the data after the project creation should be kept here.

Shhhhhubh commented on apps/project/tests/create_find_project_test.py:
```diff
@@ -0,0 +1,479 @@
+
+class TestProjectE2E(TestCase):
+    class Mutation:
+        CREATE = """
```
I already have commented on this. Please fix.

Shhhhhubh commented on apps/project/tests/create_find_project_test.py:
```diff
@@ -0,0 +1,479 @@
+                status
+                processingStatus
+                progress
+                tutorialId
+                tutorial {
+                  id
+                }
```
This is not even defined when creating a project. Why fetch this data?

Shhhhhubh commented on apps/project/tests/create_find_project_test.py:
```diff
@@ -0,0 +1,479 @@
+                  }
+                }
+              }
+            }
+          }
+        }
+    """
+        UPDATE_PROCESSED_PROJECT = """
+            mutation UpdateProcessedProject($pk: ID!, $data: ProcessedProjectUpdateInput!) {
+            updateProcessedProject(pk: $pk, data: $data) {
+                ... on OperationInfo {
+                __typename
+                messages {
```
Please format this.

Shhhhhubh commented on apps/project/tests/create_find_project_test.py:
```diff
@@ -0,0 +1,479 @@
+            }
+        """
+
+    @staticmethod
+    def remove_ignored_keys(obj, keys_to_ignore):
+        """Recursively remove keys from dicts if the key is in keys_to_ignore.
+        Works on nested dicts and lists.
+        """
+        if isinstance(obj, dict):
+            for key in list(obj.keys()):
+                if key in keys_to_ignore:
+                    obj.pop(key)
+                else:
+                    TestProjectE2E.remove_ignored_keys(obj[key], keys_to_ignore)
+        elif isinstance(obj, list):
+            for item in obj:
+                TestProjectE2E.remove_ignored_keys(item, keys_to_ignore)
+        return obj
```
Move this to utils.

Shhhhhubh commented on apps/project/tests/create_find_project_test.py:
```diff
@@ -0,0 +1,479 @@
+        cls.image_filename = Path(settings.BASE_DIR) / "apps/project/asset/image.jpg"
+        cls.aoi_geometry_filename = Path(settings.BASE_DIR) / "apps/project/asset/find_geojson.geojson"
+
+    def load_inputs(self) -> dict[str, typing.Any]:
+        with self.filename.open("r", encoding="utf-8") as f:
+            return json.load(f)
+
```
Don't need this function.

Shhhhhubh commented on apps/project/tests/create_find_project_test.py:
```diff
@@ -0,0 +1,479 @@
+        update_project_status_1_resp = update_project_status_1["data"]["updateProjectStatus"]
+        assert update_project_status_1_resp is not None, "updateProjectStatus GraphQL response is None"
+
+        # Use factory for tutorial creation.
```
Add a TODO that we need to use serializers instead of factory.

Shhhhhubh commented on apps/project/tests/create_find_project_test.py:
```diff
@@ -0,0 +1,479 @@
+        update_processed_result = update_processed_resp["result"]
+
+        # Compare key fields after processed update.
+        assert {
+            update_processed_result["projectInstruction"],
+        } == {
+            update_processed_project_data["projectInstruction"],
+        }
```
These kind of tests will automatically be covered.

Shhhhhubh commented on apps/project/tests/create_find_project_test.py:
```diff
@@ -0,0 +1,479 @@
+        update_project_status_2_resp = update_project_status_2["data"]["updateProjectStatus"]
+        assert update_project_status_2_resp is not None, "updateProjectStatus GraphQL response is None"
+
+        # Comparing project name.
+        processed_project = Project.objects.get(id=update_processed_result["id"])
+        assert update_processed_result["name"] == processed_project.generate_name()
+
+        # Firebase validation for Project.
+        project_fb_id = Project.objects.get(id=result["id"]).firebase_id
```
Do not directly use models.
Using these break our black box model of testing.

Shhhhhubh commented on apps/project/tests/create_find_project_test.py:
```diff
@@ -0,0 +1,479 @@
+        # Firebase validation for Project.
+        project_fb_id = Project.objects.get(id=result["id"]).firebase_id
+        project_ref = self.firebase_helper.ref(
+            Config.FirebaseKeys.project(project_fb_id),
```
Directly use the path.

Shhhhhubh commented on apps/project/tests/create_find_project_test.py:
```diff
@@ -0,0 +1,479 @@
+
+        assert project_fb_data is not None, "Firebase data is None for Project"
+        assert isinstance(project_fb_data, dict), "Firebase data is not a dictionary"
+        assert project_fb_data["created"] is not None, "Created Date is empty"
+        assert project_fb_data["projectId"] is not None, "ProjectId is empty"
+        assert project_fb_data["tutorialId"] is not None, "TutorialId is empty"
```
created should be a valid timestamp (iso standard)
projectId should be the same as the firebaseId
tutorialId should be the same as the firebaseId

Shhhhhubh commented on apps/project/tests/create_find_project_test.py:
```diff
@@ -0,0 +1,479 @@
+        filtered_project_expected = self.remove_ignored_keys(data["final_processed_project"], ignored_project_keys)
+
+        assert filtered_project_actual == filtered_project_expected, (
+            "Differences found when comparing Firebase data to fixed file create_find_project.json for project data."
```
"Difference found for project data in firebase"

tnagorra reviewed (changes requested):
Let's fix all the comments.

Shhhhhubh committed:
test(project): Add test files for compare project type

tnagorra committed:
test(project): merge e2e test for compare and find

- refactor e2e test
    - add test for dynamic values: projectId, created, createdBy, etc.
- move pytest to dev dependency

tnagorra committed:
fix(rastertile): addd mkt param for bing url

tnagorra committed:
fix(project): update value fo requiredCount in project in firebase

tnagorra committed:
fix(project): use enum label on error message instead of enum value
